File: 039_combination_sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int> > result;
        vector<int> path;
        sort(candidates.begin(), candidates.end());
        dfs(candidates,0,target,0,path,result);
        return result;
    }
    void dfs(vector<int>& candidates, int sum, int target, int start, vector<int>&path, vector<vector<int> > & result) {
        if (sum>target) return;
        if (sum==target) result.push_back(path);
        for (int i=start;i<candidates.size();i++) {
            if (sum+candidates[i]>target) break;
            path.push_back(candidates[i]);
            dfs(candidates, sum+candidates[i],target, i, path,result);
            path.pop_back();
        }
        return;
    }
};

File: algorithms/python3/039_combination_sum.py
class Solution:
    def combinationSum(self, candidates, target):
        candidates.sort()
        res=[]

        def dfs(target,start,path):
            if target==0:
                res.append(path)
                return
            if target<0:
                return
            for i in range(start,len(candidates)):
                if candidates[i]>target:
                    break
                dfs(target-candidates[i],i,path+[candidates[i]])
        dfs(target,0,[])
        return res


a=Solution()
nums=[2,3,6,7]


print(a.combinationSum(nums,7))

